iree/compiler/Codegen/SPIRV/MaliConfig.cpp

```diff
@@ -173,141 +173,23 @@ static LogicalResult setOpConfig(linalg::MatmulOp op) {
-  ArrayRef<int64_t> inputShape = getUntiledShape(op.inputs()[0]);
-  ArrayRef<int64_t> outputShape = getUntiledResultShape(linalgOp, 0);
-
-  if (llvm::any_of(inputShape, ShapedType::isDynamic) ||
-      llvm::any_of(outputShape, ShapedType::isDynamic)) {
-    return success();
-  }
-
-  // A list of preferred tile sizes and workgroup sizes.
-  // TODO(antiagainst): This is for Valhall now; need to consider other Mali
-  // architectures.
-  static const TileWorkgroupSizePair tileWorkgroupSizePairs[] = {
-      {{2, 2, 32}, {8, 2, 2}},
-      {{1, 4, 16}, {4, 4, 1}},
-      {{1, 1, 64}, {16, 1, 1}},
-      {{4, 4, 8}, {2, 4, 2}},
-  };
-
-  for (const auto &pair : tileWorkgroupSizePairs) {
-    const std::array<int64_t, 3> &tileSize = pair.tileSize;
-    const std::array<int64_t, 3> &workgroupSize = pair.workgroupSize;
-
-    bool isOutputTilable = outputShape[0] == 1 &&
-                           (outputShape[1] % tileSize[0] == 0) &&
-                           (outputShape[2] % tileSize[1] == 0) &&
-                           (outputShape[3] % tileSize[2] == 0);
-    if (!isOutputTilable) continue;
-
-    auto pipeline = IREE::HAL::DispatchLoweringPassPipeline::SPIRVVectorize;
-
-    TileSizesListType tileSizes;
-    // Workgroup level.
-    tileSizes.push_back({/*batch=*/0,
-                         /*output_height=*/tileSize[0],
-                         /*output_width=*/tileSize[1],
-                         /*output_channel=*/tileSize[2]});
-
-    // No tiling at the subgroup level given that we don't use subgroup
-    // level syncrhonization  or shared memory.
-    tileSizes.emplace_back();
-    tileSizes.push_back({/*batch=*/0,
-                         /*output_height=*/tileSize[0] / workgroupSize[2],
-                         /*output_width=*/tileSize[1] / workgroupSize[1],
-                         /*output_channel=*/tileSize[2] / workgroupSize[0]});
-
-    auto funcOp = op->getParentOfType<FuncOp>();
-    if (failed(setOpConfigAndEntryPointFnTranslation(
-            funcOp, op, tileSizes, {}, pipeline, workgroupSize))) {
-      return failure();
-    }
-
-    return defineConvWorkgroupCountRegion(
-        op, llvm::makeArrayRef(outputShape).drop_front(), tileSize);
-  }
-  return success();
-}
-
 //===----------------------------------------------------------------------===//
 // Entry Point
 //===----------------------------------------------------------------------===//
 
 LogicalResult setMaliCodeGenConfig(const spirv::TargetEnv &,
                                    Operation *rootOp) {
   return TypeSwitch<Operation *, LogicalResult>(rootOp)
-      .Case<linalg::BatchMatmulOp, linalg::Conv2DNhwcHwcfOp,
-            linalg::DepthwiseConv2DNhwOp, linalg::MatmulOp>(
+      .Case<linalg::BatchMatmulOp, linalg::MatmulOp>(
           [](auto op) { return setOpConfig(op); })
+      .Case<linalg::Conv2DNhwcHwcfOp>([](auto op) {
+        return setConvOpConfig(op, /*subgroupSize=*/16,
```

Comment:
This has a compile time overhead. Each `Case` is evaluated sequentially. Using the overloaded methods means, once resolved it will dispatch to the right function with no overhead.

Reply:
Not sure I follow exactly. You mean compile time or run time overhead? `TypeSwitch` under the hood are translated  into `if (... dyn_cast ...)` chain so run time it should be similar there. For compilation time I'm not quite follow where the overhead come from. Even if you put multiple ops in the same `Case`, [under the hood](https://github.com/llvm/llvm-project/blob/d480f968ad8b56d3ee4a6b6df5532d485b0ad01e/llvm/include/llvm/ADT/TypeSwitch.h#L35-L42) it still expands to multiple `Case`s for each of them.